clash-cores/test/Test/Cores/LineCoding8b10b.hs

```diff
@@ -0,0 +1,115 @@
+-- |
+--   Copyright   :  (C) 2024, QBayLogic B.V.
+--   License     :  BSD2 (see the file LICENSE)
+--   Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+--
+--   8b/10b encoding and decoding tests
+module Test.Cores.LineCoding8b10b where
+
+import Clash.Cores.LineCoding8b10b
+import Clash.Hedgehog.Sized.BitVector
+import qualified Clash.Prelude as C
+import Data.Maybe (isNothing)
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import Test.Tasty
+import Test.Tasty.Hedgehog
+import Test.Tasty.TH
+import Prelude
+
+-- | Check if a 'BitVector' does not contain a sequence of bits with the same
+--   value for 5 or more indices consecutively
+checkBitSequence :: C.BitVector 10 -> Bool
+checkBitSequence cg =
+  isNothing $
+    C.elemIndex True $
+      C.map (f . C.pack) $
+        C.windows1d C.d5 $
+          C.bv2v cg
+ where
+  f a = a == 0b11111 || a == 0b00000
+
+-- | Function that creates a list with a given range that contains a list of
+--   running disparities and 'Symbol8b10b's
+genSymbol8b10bs :: H.Range Int -> H.Gen [(Bool, Symbol8b10b)]
+genSymbol8b10bs range = do
+  n <- Gen.int range
+  genSymbol8b10bs1 n False
+
+-- | Recursive function to generate a list of 'Symbol8b10b's with the correct
+--   running disparity
+genSymbol8b10bs1 :: Int -> Bool -> H.Gen [(Bool, Symbol8b10b)]
+genSymbol8b10bs1 0 _ = pure []
+genSymbol8b10bs1 n rd = do
+  (rdNew, dw) <- genSymbol8b10b rd
+  ((rdNew, dw) :) <$> genSymbol8b10bs1 (pred n) rdNew
+
+-- | Generate a 'Symbol8b10b' by creating a 'BitVector' of length 10 and
+--   decoding it with the 'decode8b10b' function
+genSymbol8b10b :: Bool -> H.Gen (Bool, Symbol8b10b)
+genSymbol8b10b rd = Gen.filter f $ decode8b10b rd <$> genDefinedBitVector
+ where
+  f (_, dw) = isDw dw
+
+-- Check if the output of 'decode8b10b' is a valid value for a given value from
+-- 'encode8b10b'.
+prop_decode8b10bCheckNothing :: H.Property
+prop_decode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = isValidSymbol dw1 && isValidSymbol dw2
+       where
+        (_, dw1) = decode8b10b False $ snd $ encode8b10b False (Dw inp)
+        (_, dw2) = decode8b10b True $ snd $ encode8b10b True (Dw inp)
+
+  H.assert out
+
+-- | Encode the input signal and check whether it is a valid value. It should be
+--   valid for every possible input.
+prop_encode8b10bCheckNothing :: H.Property
+prop_encode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = 0 /= snd (encode8b10b False (Dw inp))
+
+  H.assert out
```

Comment:
```suggestion
  snd (encode8b10b False (Dw inp)) H./== 0
```

I don't think the doc for this function really matches what it does, though. It just checks that it never produces a 10-bit zero. There's a whole bunch of other invalid codes that are not checked.

Reply:
Hmmm, that might be true. I will also take a look at this tomorrow.

Reply:
Ah, got it, it's because of this line line in `encode8b10b`:

`out = if isValidSymbol sym then (rdNew, cg) else (rd, 0)`

So if `sym` is not a valid symbol, the output will always be 0. The description should then be:

```
-- | Encode the input signal and check whether it is a valid value. It should be
--   valid for every valid input symbol.
```

Reply:
And of course in this case every input value is indeed valid, as it's hard-coded to be a `Dw`.

Reply:
(We have decided to scrap this test as it does not really make a lot of sense)